### ISSUE TITLE
Compiling list file

### ISSUE THREAD

**Issue by Shiska** [Zeex/pawn#216]
After looking through pc_compile it shouldn't be to difficult to implement (I guess) and it would be helpful because I want to add an additional step between preprocessing and compilation with external code

**Comment by Daniel-Cortez** [comment]
You mean the `-l` option?

**Comment by Shiska** [comment]
Well the -l option only generates the list file but these is no way to compile the list file to an amx afterwards

**Comment by Y-Less** [comment]
Unfortunately, currently `-l` output is not totally accurate.  Or more correctly, it isn't the FINAL version of code.  Take this example:

```pawn
Func1()
{
    #if defined Func2
        Func2();
    #endif
}

Func2()
{
    printf("hi");
}

main()
{
    Func1();
}
```

Running this will output:

```
hi
```

Compiling this with `-l` will give:

```pawn
Func1()
{
}

Func2()
{
    printf("hi");
}

main()
{
    Func1();
}
```

Which if you were then able to compile, would NOT output `hi`.  The reason for this is the PAWN compiler's multi-pass compilation.  The first pass generates some code, and collects a list of all the functions and other symbols that exist.  During this first pass, while `Func1` is being compiled the compiler does not yet know that `Func2` exists, so the `#if` fails.  After this pass is when `-l` is dumped.

The second pass uses the list of functions gathered in the first pass from the start, so at the time of compiling `Func1`, the compiler already knows that `Func2` will exist in the future.  Side note, you can really upset the compiler with this if you have functions that exist in the first pass not exist in the second pass, but I won't go in to that.  In theory, after this pass would be the better time to dump `-l` output, but I don't think the original version did as they didn't expect odd situations like this.

And this is not an obscure corner-case - it is a corner-case, but it is also the fundamental operation on which ALS, y_groups, and probably more code are all based.  So if you use almost any library at all ever, callback chaining won't work if you try and compile from `-l` output, at least without adjusting how that is generated first.

**Comment by Shiska** [comment]
Ah yeah, forgot about that completely but I see no problem adjusting the `-l` option so it dumps after the second pass, otherwise the list file would be useless, not that anyone used it

Than only setfiledirect needs to be fixed, #file likes to appear at the end of a line, probably if the source file has no new line at the end and that the quotes are missing around the filename
#file is also used if there are more than one sourcefile in pc_compile (didn't even know that this is possible)

After that it should compile without problems

**Comment by Y-Less** [comment]
The only thing to watch out for then is static variables in includes.  Here is a horrible test case for you:

other.inc

```pawn
#if defined _INC_other
	#endinput
#endif
#define _INC_other

#if NUMBER == 1
	#define NUM_BACKUP 1
#elseif NUMBER == 2
	#define NUM_BACKUP 2
#endif

#undef NUMBER
#undef FUNCTION

#define NUMBER (3)
#define FUNCTION Func3

#include <include>

#if NUM_BACKUP == 1
	#define NUMBER 1
	#define FUNCTION Func1
#elseif NUM_BACKUP == 2
	#define NUMBER 2
	#define FUNCTION Func2
#endif
```

include.inc:

```pawn
#if defined _inc_include
	#undef _inc_include
#endif

static gVar = NUMBER;

#include <other>

FUNCTION()
{
	return gVar;
}

#undef FUNCTION
#undef NUMBER
```

listfile.pwn:

```pawn
native printf(const format[], {Float,_}:...);

#define NUMBER (1)
#define FUNCTION Func1
#include <include>

#define NUMBER (2)
#define FUNCTION Func2
#include <include>

main()
{
	printf("%d %d %d", Func1(), Func2(), Func3());
}
```

That will print `1 2 3`, but compiled with `-l` will give:

```pawn
#pragma ctrlchar 0x5c
#pragma pack false
#pragma semicolon true
#pragma tabsize 8

#file "D:\work\SA-MP\gamemodes\listfile.pwn"
#line 1
native printf(const format[], {Float,_}:...);


#file "D:\work\SA-MP\pawno\include\include.inc"
#line 4

static gVar = (1);


#file "D:\work\SA-MP\pawno\include\other.inc"
#line 5

#line 11

#line 14

#line 17


#file "D:\work\SA-MP\pawno\include\include.inc"
#line 4

static gVar = (3);


#file "D:\work\SA-MP\pawno\include\other.inc"
#line 2
	#endinput

#file "D:\work\SA-MP\pawno\include\include.inc"
#line 8

Func3()
{
	return gVar;
}

#line 16



#file "D:\work\SA-MP\pawno\include\other.inc"
#line 19

#line 27



#file "D:\work\SA-MP\pawno\include\include.inc"
#line 8

Func1()
{
	return gVar;
}

#line 16



#file "D:\work\SA-MP\gamemodes\listfile.pwn"
#line 6


#file "D:\work\SA-MP\pawno\include\include.inc"
#line 4

static gVar = (2);


#file "D:\work\SA-MP\pawno\include\other.inc"
#line 2
	#endinput

#file "D:\work\SA-MP\pawno\include\include.inc"
#line 8

Func2()
{
	return gVar;
}

#line 16



#file "D:\work\SA-MP\gamemodes\listfile.pwn"
#line 10

main()
{
	printf("%d %d %d", Func1(), Func2(), Func3());
}
```

Figuring out variable scopes from that is non-trivial, there's no way to tell when `#file include.inc` appears because it was included, or because its included files ended.  And thus there's no way to easilly tell which `gVar` goes with which `FuncX`.  If you can figure that out in a generic way, then you're sorted.

**Comment by Y-Less** [comment]
There is already `#section`, which could be useful.  But you'd need a way to support nested sections and thus explicitly end sections.

**Comment by Shiska** [comment]
Didn't expect to encounter that much problems but its halfway important to get one complete file so I will try to figure something out (I hope) and be back :)

Thanks for the help

**Comment by Daniel-Cortez** [comment]
Now, after #375 is merged, the list file is generated after the second pass and should be compilable, shouldn't it?

**Comment by Southclaws** [comment]
Yeah, it should. It may need some tests though, for possible corner cases.
